lychee-lib/src/uri.rs

```diff
@@ -1,200 +1,222 @@
     }
 
     #[must_use]
     pub fn host_ip(&self) -> Option<IpAddr> {
-        match self {
-            Self::Website(url) => match url.host()? {
-                url::Host::Ipv4(v4_addr) => Some(v4_addr.into()),
-                url::Host::Ipv6(v6_addr) => Some(v6_addr.into()),
-                url::Host::Domain(_) => None,
-            },
-            Self::Mail(_) => None,
+        match self.url.host()? {
+            url::Host::Domain(_) => None,
+            url::Host::Ipv4(v4_addr) => Some(v4_addr.into()),
+            url::Host::Ipv6(v6_addr) => Some(v6_addr.into()),
         }
     }
 
+    // TODO: Support GitLab etc.
     pub(crate) fn extract_github(&self) -> Option<(&str, &str)> {
         debug_assert!(
-            matches!(self, Uri::Website(_)),
+            !matches!(self.scheme(), "mailto"),
             "Should only be called on a Website type!"
         );
-        let url = if let Uri::Website(url) = self {
-            url
-        } else {
-            unsafe { std::hint::unreachable_unchecked() }
-        };
-        if matches!(url.domain(), Some("github.com") | Some("www.github.com")) {
-            if let Some(mut path) = url.path_segments() {
-                if let Some(owner) = path.next() {
-                    if let Some(repo) = path.next() {
-                        return Some((owner, repo));
-                    }
-                }
-            }
+
+        // TODO: Support more patterns
+        if matches!(
+            self.domain()?,
+            "github.com" | "www.github.com" | "raw.githubusercontent.com"
+        ) {
+            let mut path = self.path_segments()?;
+            let owner = path.next()?;
+            let repo = path.next()?;
+            return Some((owner, repo));
         }
+
         None
     }
 }
 
-fn is_internal_link(link: &str) -> bool {
-    // The first element should contain the Markdown file link
-    // @see https://www.markdownguide.org/basic-syntax/#links
-    let anchor_links = link.split('#').next().unwrap_or("");
-    anchor_links.ends_with(".md") | anchor_links.ends_with(".markdown")
-}
-
 impl AsRef<str> for Uri {
     fn as_ref(&self) -> &str {
         self.as_str()
     }
 }
 
+impl From<Url> for Uri {
+    fn from(url: Url) -> Self {
+        Self { url }
+    }
+}
+
+impl TryFrom<String> for Uri {
+    type Error = ErrorKind;
+
+    fn try_from(s: String) -> Result<Self> {
+        let s = s.trim_start_matches("mailto:");
+        if let Err(mail_err) = parse_email(s) {
+            match Url::parse(s) {
+                Ok(uri) => Ok(uri.into()),
+                Err(url_err) => Err((s.to_owned(), url_err, mail_err).into()),
+            }
+        } else {
+            Ok(Url::parse(&(String::from("mailto:") + s)).unwrap().into())
+        }
+    }
+}
+
 impl TryFrom<&str> for Uri {
     type Error = ErrorKind;
 
     fn try_from(s: &str) -> Result<Self> {
-        // Check for internal Markdown links
-        let is_link_internal = is_internal_link(s);
```

Comment:
Doesn't that mean that relative links in Markdown won't work anymore?
https://github.com/lycheeverse/lychee/issues/146
Or is that covered by your refactoring somehow?

Reply:
Tests are passing, so it seems to be fine.

Reply:
Yep I should mention this.
I don't get the point of `ignore` that test.
What's the background for this check?